matrix_content_scanner/logutils.py

```diff
@@ -0,0 +1,65 @@
+#  Copyright 2022 The Matrix.org Foundation C.I.C.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+import logging
+from contextvars import ContextVar
+from typing import Any, Callable
+
+from twisted.web.http import Request
+
+# The serverName/mediaId path of the media.
+media_path: ContextVar[str] = ContextVar("media_path")
+# The request being performed (download, thumbnail, scan, etc).
+request_type: ContextVar[str] = ContextVar("request_type")
+
+
+def get_factory() -> Callable[..., logging.LogRecord]:
+    old_factory = logging.getLogRecordFactory()
+
+    def _factory(*args: Any, **kwargs: Any) -> logging.LogRecord:
+        record = old_factory(*args, **kwargs)
+        # Define custom attributes on the records. We need to ignore the types here
+        # because otherwise mypy complains the attributes aren't defined on LogRecord.
+        record.media_path = media_path.get(None)  # type: ignore[attr-defined]
+        record.request_type = request_type.get(None)  # type: ignore[attr-defined]
+        return record
+
+    return _factory
+
+
+def set_context_from_request(request: Request) -> None:
+    """Set the media_path and request_type ContextVars from the given request if possible.
+
+    Args:
+        request: The request to set the context from.
+    """
+    assert request.path is not None
+    path = request.path.decode("utf-8")
+
+    # We're only interested in the bit after /_matrix/media_proxy/unstable
+    parts = path.split("/")[4:]
+    request_type.set(parts[0])
+
+    # If we have more than one part, then we likely have the media path as well in the
+    # request's path.
+    if len(parts) == 3:
```

Comment:
this reels really quite magical. Why exactly 3?

Reply:
I don't really feel like this has been answered - and more to the point, it's different to what the comment on the line above says (2 and 4 are also greater than 1, but are not covered here)

Looking at https://github.com/matrix-org/matrix-content-scanner-python/blob/f2dd6f978a06537254f66819caa070840dab6f30/docs/api.md, I see we're trying to catch `/download`, `/thumbnail` and `/scan`, each of which has three parts. Wouldn't it be easier/more reliable just to call `set_media_path_in_context` from `DownloadServlet`, `ThumbnailServlet`, `ScanServlet`?

Reply:
Yeah probably, especially since we already need to parse the media path in these servlets.